src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp

```diff
@@ -49,132 +50,197 @@ HypreExport::~HypreExport()
 {
-  if( m_targetRank >= 0 )
+  if( m_subComm != MPI_COMM_NULL )
   {
-    int const rank = MpiWrapper::commRank( vec.getComm() );
-    GEOSX_ERROR_IF( rank == m_targetRank && !values, "HypreExport: must pass non-null pointers on the target rank" );
+    hypre_Vector * localVector{};
+    if( MpiWrapper::commRank( vec.getComm() ) == m_targetRank )
+    {
+      GEOSX_LAI_ASSERT_EQ( values.size(), vec.globalSize() );
+      values.move( LvArray::MemorySpace::host, false );
 
-    if( m_subComm != MPI_COMM_NULL )
+      // HACK: create a hypre vector that points to local data; we have to use const_cast,
+      //       but this is ok because we don't modify the values, only scatter the vector.
+      localVector = hypre_SeqVectorCreate( LvArray::integerConversion< HYPRE_Int >( values.size() ) );
+      hypre_VectorOwnsData( localVector ) = false;
+      hypre_VectorData( localVector ) = const_cast< real64 * >( values.data() );
+      hypre_SeqVectorInitialize_v2( localVector, HYPRE_MEMORY_HOST );
+    }
+
+    // scatter the data
+    hypre_ParVector * const parVector = hypre_VectorToParVector( m_subComm,
+                                                                 localVector,
+                                                                 hypre_ParVectorPartitioning( vec.unwrapped() ) );
+    // copy local part of the data over to the output vector
+    HYPRE_Real const * const parVectorData = hypre_VectorData( hypre_ParVectorLocalVector( parVector ) );
+    if( hypre_ParVectorMemoryLocation( vec.unwrapped() ) == HYPRE_MEMORY_HOST )
     {
-      HYPRE_Vector localVector = nullptr;
-      if( rank == m_targetRank )
-      {
-        // HACK: create a hypre vector that points to local data; we have to use const_cast,
-        //       but this is ok because we don't modify the values, only scatter the vector.
-        localVector = HYPRE_VectorCreate( LvArray::integerConversion< HYPRE_Int >( vec.globalSize() ) );
-        hypre_VectorOwnsData( ( hypre_Vector * ) localVector ) = false;
-        hypre_VectorData( ( hypre_Vector * ) localVector ) = const_cast< real64 * >( values );
-        hypre_SeqVectorInitialize_v2( ( hypre_Vector * ) localVector, HYPRE_MEMORY_HOST );
-      }
-
-      // output vector partitioning array
-      HYPRE_BigInt * const partitioning = hypre_ParVectorPartitioning( vec.unwrapped() );
-
-      // scatter the data
-      HYPRE_ParVector parVector;
-      GEOSX_LAI_CHECK_ERROR( HYPRE_VectorToParVector( m_subComm,
-                                                      localVector,
-                                                      partitioning,
-                                                      &parVector ) );
-
-
-      // copy parVector local part over to the output vector
-      HYPRE_Real * const parVectorData = hypre_VectorData( hypre_ParVectorLocalVector( parVector ) );
       std::copy( parVectorData, parVectorData + vec.localSize(), vec.extractLocalVector() );
```

Comment:
```suggestion
      copyOrTransform( parVectorData, parVectorData + vec.localSize(), vec.extractLocalVector() );
```

Reply:
Sure, why not. In this case types are guaranteed to be the same by `static_assert` so it should always call the copy version, but for uniformity we can do this.